dbms/src/Common/Exception.h

```diff
@@ -172,28 +162,98 @@ std::enable_if_t<std::is_pointer_v<T>, T> exception_cast(std::exception_ptr e)
 
 namespace exception_details
 {
+inline std::string generateFormattedMessage(const char * condition)
+{
+    return fmt::format("Assert {} fail!", condition);
+}
+
 template <typename T, typename... Args>
-inline std::string generateLogMessage(const char * condition, T && fmt_str, Args &&... args)
+inline std::string generateFormattedMessage(const char * condition, T && fmt_str, Args &&... args)
+{
+    return FmtBuffer().fmtAppend("Assert {} fail! ", condition).fmtAppend(fmt_str, std::forward<Args>(args)...).toString();
+}
+
+template <typename... Args>
+inline Poco::Message generateLogMessage(const std::string & logger_name, const char * filename, int lineno, const char * condition, Args &&... args)
+{
+    return Poco::Message(
+        logger_name,
+        generateFormattedMessage(condition, std::forward<Args>(args)...),
+        Poco::Message::PRIO_FATAL,
+        filename,
+        lineno);
+}
+
+const LoggerPtr & getDefaultFatalLogger();
+
+inline void log(const char * filename, int lineno, const char * condition, const LoggerPtr & logger)
+{
+    if (likely(logger->fatal()))
+    {
+        auto message = generateLogMessage(logger->name(), filename, lineno, condition);
+        logger->log(message);
+    }
+}
+
+inline void log(const char * filename, int lineno, const char * condition)
+{
+    log(filename, lineno, condition, getDefaultFatalLogger());
+}
+
+template <typename... Args>
+inline void log(const char * filename, int lineno, const char * condition, const LoggerPtr & logger, const char * fmt_str, Args &&... args)
+{
+    if (logger->fatal())
+    {
+        auto message = generateLogMessage(
+            logger->name(),
+            filename,
+            lineno,
+            condition,
+            fmt_str,
+            std::forward<Args>(args)...);
+        logger->log(message);
+    }
+}
+
+template <typename... Args>
+inline void log(const char * filename, int lineno, const char * condition, const char * fmt_str, Args &&... args)
```

Comment:
ditto.

Reply:
this `fmt_str` is used for overloadding from `const LoggerPtr &` or this version will always be chosen even when there's a `const LoggerPtr &`.